middlewares/authenticator_test.go

```diff
@@ -5,13 +5,52 @@ import (
 	"io/ioutil"
 	"net/http"
 	"net/http/httptest"
+	"os"
 	"testing"
 
 	"github.com/codegangsta/negroni"
 	"github.com/containous/traefik/types"
 	"github.com/stretchr/testify/assert"
 )
 
+func TestBasicAuthUsersFromFile(t *testing.T) {
+	usersStr := "test:$apr1$H6uskkkW$IgXLP6ewTrSuBkTrqE8wj/\ntest2:$apr1$d9hr9HBB$4HxwgUir3HP4EsggP/QNo0\n"
+	usersFile, err := ioutil.TempFile("", "basic-users")
+	assert.NoError(t, err, "there should be no error")
+	defer os.Remove(usersFile.Name())
+	_, err = usersFile.Write([]byte(usersStr))
+	assert.NoError(t, err, "there should be no error")
+	basic := &types.Basic{
+		UsersFile: usersFile.Name(),
+	}
+	users, err := parserBasicUsers(basic)
+	assert.NoError(t, err, "there should be no error")
+	assert.Equal(t, 2, len(users), "they should be equal")
+	_, ok := users["test"]
+	assert.True(t, ok, "user test should be found")
+	_, ok = users["test2"]
+	assert.True(t, ok, "user test2 should be found")
+}
+
+func TestDigestAuthUsersFromFile(t *testing.T) {
+	usersStr := "test:traefik:a2688e031edb4be6a3797f3882655c05 \ntest2:traefik:518845800f9e2bfb1f1f740ec24f074e\n"
+	usersFile, err := ioutil.TempFile("", "basic-users")
+	assert.NoError(t, err, "there should be no error")
+	defer os.Remove(usersFile.Name())
+	_, err = usersFile.Write([]byte(usersStr))
+	assert.NoError(t, err, "there should be no error")
+	digest := &types.Digest{
+		UsersFile: usersFile.Name(),
+	}
+	users, err := parserDigestUsers(digest)
+	assert.NoError(t, err, "there should be no error")
+	assert.Equal(t, 2, len(users), "they should be equal")
+	_, ok := users["test:traefik"]
+	assert.True(t, ok, "user test should be found")
+	_, ok = users["test:traefik"]
+	assert.True(t, ok, "user test2 should be found")
+}
```

Comment:
To me, the decision to use a table-driven test or not is determined by the trade-off to be made between the amount of variance that needs to go into the test and the level of duplication you can remove. The two tests do seem to have quite a bit of code in common.

For the sake of comparability, here are the new tests in a table-driven fashion:

```go
func TestAuthUsersFromFile(t *testing.T) {
	tests := []struct {
		authType   string
		usersStr   string
		userKeys   []string
		parserFunc func(fileName string) (map[string]string, error)
	}{
		{
			authType: "basic",
			usersStr: "test:$apr1$H6uskkkW$IgXLP6ewTrSuBkTrqE8wj/\ntest2:$apr1$d9hr9HBB$4HxwgUir3HP4EsggP/QNo0\n",
			userKeys: []string{"test", "test2"},
			parserFunc: func(fileName string) (map[string]string, error) {
				basic := &types.Basic{
					UsersFile: fileName,
				}
				return parserBasicUsers(basic)
			},
		},
		{
			authType: "digest",
			usersStr: "test:traefik:a2688e031edb4be6a3797f3882655c05 \ntest2:traefik:518845800f9e2bfb1f1f740ec24f074e\n",
			userKeys: []string{"test:traefik", "test2:traefik"},
			parserFunc: func(fileName string) (map[string]string, error) {
				digest := &types.Digest{
					UsersFile: fileName,
				}
				return parserDigestUsers(digest)
			},
		},
	}

	for _, test := range tests {
		test := test
		t.Run(test.authType, func(t *testing.T) {
			t.Parallel()
			usersFile, err := ioutil.TempFile("", "auth-users")
			assert.NoError(t, err, "there should be no error")
			defer os.Remove(usersFile.Name())
			_, err = usersFile.Write([]byte(test.usersStr))
			assert.NoError(t, err, "there should be no error")
			users, err := test.parserFunc(usersFile.Name())
			assert.NoError(t, err, "there should be no error")
			assert.Equal(t, 2, len(users), "they should be equal")
			_, ok := users[test.userKeys[0]]
			assert.True(t, ok, "user test should be found")
			_, ok = users[test.userKeys[1]]
			assert.True(t, ok, "user test2 should be found")
		})
	}
}
```

For me, the only thing that tips this from a "pro" table-driven to a "maybe" is, again, the distinct authentication types. If those where the same, `parserFunc` could be directly assigned to the respective function under test.

Personally, I prefer the table-driven approach in this case. I'm fine with sticking with two tests as well though if you believe that's better.